<h1 align="center">Exploratory Data Analysis (EDA)</h1>
<h1 align="center">Explorative Datenanalyse</h1>


In [20]:
# libraries
import os
import pandas as pd
import duckdb
import numpy as np
import plotly.express as px
import matplotlib.pyplot as plt
import requests


# Paths
BASE_DIR = os.path.dirname(os.path.abspath('__file__'))
DATA_DIR = os.path.join(BASE_DIR, 'data')
WEATHER_DIR = '/Users/lamiaabahgat/Documents/Portfolio/data_4_years/pycharm/venv/weather_data_output'


In [2]:
#Loading file
# Datei wird geladen

df = pd.read_csv(os.path.join(DATA_DIR, 'cleaned_car_dataset.csv'), low_memory=False)


In [25]:
df.head()

,state,government_district,district,municipality,official_admin_code,accident_uid,year,month,hour,day,...,bike_involved,passenger_involved,other_involved,utm_easting,utm_northing,longitude,latitude,month_num,estimated_date,datetime
0,schleswig-holstein,none,nordfriesland,niebüll,1054165,01210308125013512021,2021,March,7,monday,...,acc_nobike,no_pass,no_other,483995.394384,6.069091e+06,8.751233,54.768787,3,2021-03-01,2021-03-01 07:00:00
1,schleswig-holstein,none,"kiel, landeshauptstadt",kiel,1002000,01210608134013112021,2021,June,15,tuesday,...,acc_nobike,no_pass,no_other,573010.098061,6.020091e+06,10.122558,54.323450,6,2021-06-01,2021-06-01 15:00:00
2,schleswig-holstein,none,steinburg,wilster,1061007,01210610181013902021,2021,June,13,thursday,...,acc_bike,no_pass,other,527231.388501,5.972658e+06,9.414457,53.901645,6,2021-06-03,2021-06-03 13:00:00
3,schleswig-holstein,none,herzogtum lauenburg,ratzeburg,1053041,01210524161013132021,2021,May,11,monday,...,acc_bike,no_pass,no_other,614902.973625,5.963897e+06,10.745101,53.810913,5,2021-05-03,2021-05-03 11:00:00
4,schleswig-holstein,none,ostholstein,neustadt in holstein,1055032,01210529152013382022,2021,May,15,saturday,...,acc_bike,no_pass,no_other,617331.180263,5.996138e+06,10.794357,54.100018,5,2021-05-01,2021-05-01 15:00:00


In [4]:
## Section 1 — Basic Overview
## Abschnitt 1 — Überblick
# how many rows/ columns
# Wie viele Zeilen/Spalten
df.shape

(782642, 28)

In [5]:
# columns types
# Spaltentypen
df.dtypes

state                         str
government_district           str
district                      str
municipality                  str
official_admin_code         int64
accident_uid                  str
year                        int64
month                         str
hour                        int64
day                           str
category                      str
kind                        int64
type                        int64
light                         str
road_surface                  str
goods_vehicle_involved        str
passenger_car_involved        str
motorcycle_involved           str
bike_involved                 str
passenger_involved            str
other_involved                str
utm_easting               float64
utm_northing              float64
longitude                 float64
latitude                  float64
month_num                   int64
estimated_date                str
datetime                      str
dtype: object

In [6]:
# statistics
# Statistiken
df.describe()

,official_admin_code,year,hour,kind,type,utm_easting,utm_northing,longitude,latitude,month_num
count,7.826420e+05,782642.000000,782642.000000,782642.000000,782642.000000,782642.000000,7.826420e+05,782642.000000,782642.000000,782642.000000
mean,7.687589e+06,2022.039387,13.363906,3.825872,3.813249,550212.285847,5.632605e+06,9.716770,50.823070,6.845812
std,3.601495e+06,0.814188,4.795714,2.672812,2.178701,149890.177134,2.002909e+05,2.142948,1.797303,3.162465
min,1.001000e+06,2021.000000,0.000000,0.000000,1.000000,280699.214691,5.241009e+06,5.871498,47.315548,1.000000
25%,5.315000e+06,2021.000000,10.000000,2.000000,2.000000,429318.759489,5.461004e+06,7.993287,49.289760,4.000000
50%,8.111000e+06,2022.000000,14.000000,4.000000,3.000000,533118.857381,5.652343e+06,9.472112,50.989165,7.000000
75%,9.565000e+06,2023.000000,17.000000,5.000000,6.000000,673953.178210,5.791928e+06,11.410379,52.266320,9.000000
max,1.607705e+07,2023.000000,23.000000,9.000000,7.000000,920405.489500,6.100725e+06,15.028387,55.051843,12.000000


In [ ]:
# Missing values summary / Fehlende Werte Übersicht
# district: 110 missing (0.01%) — small rural areas not in lookup table
# municipality: 436 missing (0.06%) — same reason
# All critical analysis columns have 0 missing values ✅
#
# Fehlende Werte in district/municipality stammen aus
# kleinen ländlichen Gebieten die nicht in der ASG-Tabelle sind

df.isna().sum()

state                       0
government_district         0
district                  110
municipality              436
official_admin_code         0
accident_uid                0
year                        0
month                       0
hour                        0
day                         0
category                    0
kind                        0
type                        0
light                       0
road_surface                0
goods_vehicle_involved      0
passenger_car_involved      0
motorcycle_involved         0
bike_involved               0
passenger_involved          0
other_involved              0
utm_easting                 0
utm_northing                0
longitude                   0
latitude                    0
month_num                   0
estimated_date              0
datetime                    0
dtype: int64

In [13]:
df.columns

Index(['state', 'government_district', 'district', 'municipality',
       'official_admin_code', 'accident_uid', 'year', 'month', 'hour', 'day',
       'category', 'kind', 'type', 'light', 'road_surface',
       'goods_vehicle_involved', 'passenger_car_involved',
       'motorcycle_involved', 'bike_involved', 'passenger_involved',
       'other_involved', 'utm_easting', 'utm_northing', 'longitude',
       'latitude', 'month_num', 'estimated_date', 'datetime'],
      dtype='str')

In [3]:
# Section 2 — Accident Distribution by Year / State / Category
# Interactive Icicle chart — click to drill down into states and categories
# Abschnitt 2 — Unfallverteilung nach Jahr / Bundesland / Kategorie
# Interaktives Icicle-Diagramm — zum Aufklappen klicken

import plotly.graph_objects as go
import pandas as pd

# Group data by year, state and category
# Daten nach Jahr, Bundesland und Kategorie gruppieren
df_counts = df.groupby(['year', 'state', 'category']).size().reset_index(name='accident_count')

levels = ['category', 'state', 'year']
value_column = 'accident_count'

def build_hierarchical_dataframe(df, levels, value_column):
    df_list = []
    for i, level in enumerate(levels):
        df_tree = pd.DataFrame(columns=['id', 'parent', 'value', 'color'])
        dfg = df.groupby(levels[i:])[value_column].sum().reset_index()

        # Make unique IDs by joining all levels from current to top
        dfg['unique_id'] = dfg[levels[i:]].astype(str).agg(' | '.join, axis=1)
        df_tree['id'] = dfg['unique_id']
        df_tree['value'] = dfg[value_column]
        df_tree['color'] = dfg[value_column]

        if i < len(levels) - 1:
            dfg['parent_id'] = dfg[levels[i+1:]].astype(str).agg(' | '.join, axis=1)
            df_tree['parent'] = dfg['parent_id']
        else:
            df_tree['parent'] = 'Germany'

        df_list.append(df_tree)

    # Root node — Germany total
    # Wurzelknoten — Deutschland gesamt
    total = pd.Series(dict(
        id='Germany',
        parent='',
        value=df[value_column].sum(),
        color=df[value_column].sum()
    ), name=0)
    df_list.append(total)

    df_all_trees = pd.concat(df_list, ignore_index=True)
    return df_all_trees

df_all_trees = build_hierarchical_dataframe(df_counts, levels, value_column)

# Color scale: yellow → orange → red (low → high accidents)
# Farbskala: Gelb → Orange → Rot (wenig → viele Unfälle)
custom_colorscale = [
    [0.0, '#f1c21b'],
    [0.5, '#ff832b'],
    [1.0, '#da1e28'],
]

fig = go.Figure()

fig.add_trace(go.Icicle(
    labels=df_all_trees['id'],
    parents=df_all_trees['parent'],
    values=df_all_trees['value'],
    branchvalues='total',
    marker=dict(
        colors=df_all_trees['color'],
        colorscale=custom_colorscale,
        showscale=True,
        colorbar=dict(title='Accidents')
    ),
    hovertemplate='<b>%{label}</b><br>Accidents: %{value:,}<extra></extra>',
    maxdepth=2,
    tiling=dict(
        orientation='v',
        pad=2
    ),
    textfont=dict(size=10),
    pathbar=dict(visible=True)
))

fig.update_layout(
    title=dict(
        text='🇩🇪 Accidents by Year / State / Category<br><sup>Click to drill down | Zum Aufklappen klicken</sup>',
        x=0.5,
        xanchor='center'
    ),
    margin=dict(t=80, l=25, r=25, b=25),
    width=1400,
    height=700,
    paper_bgcolor='white',
)

fig.show()

In [4]:
# Accidents by State: 24- Hour Cycle
#Unfälle nach Bundesland im 24-Stunden-Verlauf
import plotly.express as px

# Aggregating data so it's fast
df_hour = df.groupby(['state', 'hour', 'category']).size().reset_index(name='counts')

fig = px.bar(df_hour,
             x="state",
             y="counts",
             color="category",
             animation_frame="hour", # This adds the play button!
             range_y=[0, df_hour['counts'].max() + 100],
             title="Accidents by State: 24-Hour Cycle | Unfälle nach Bundesland im 24-Stunden-Verlauf ")

fig.update_layout(xaxis={'categoryorder':'total descending','tickangle': 25})
fig.show()


In [5]:
# Accident Severity Across German States by Light Condition
# Vergleich der Unfallschwere in deutschen Bundesländern je nach Lichtverhältnissen

# Aggregating data so it's fast
# Datenaggregation zur Leistungsoptimierung
df_light = df.groupby(['light', 'state', 'category']).size().reset_index(name='accident_count')
fig = px.bar(df_light,
             x="category",
             y="accident_count",
             color="light",
             facet_col="state",      # Creates a chart for each state
             facet_col_wrap=4,       # Puts 4 charts per row
             title=" Accidents by light conditions | Unfälle nach Lichtverhältnissen")

fig.update_layout(height=800)
fig.show()

In [6]:
# Which states have most accidents?
# Welche Bundesländer haben die meisten Unfälle?

# Grouping by State and Category
df_state = df.groupby(['state', 'category']).size().reset_index(name='total')

fig = px.bar(df_state,
             x="total",
             y="state",
             color="category",
             orientation='h', # Horizontal looks better for long State names
             text_auto='.2s', # Adds shortened numbers (like 12k) on the bars
             title="Total Accidents by State and Severity | Unfälle insgesamt nach Bundesland und Schweregrad")

fig.update_layout(
    yaxis={'categoryorder':'total ascending'},
    hovermode="y unified" # Shows all category data when you hover over a state
)
fig.show()

In [7]:
# Accidents by day of week
# Unfälle nach Wochentag

# Grouping by Day and State
# Gruppierung nach Tag und Bundesland
df_day = df.groupby(['day', 'state']).size().reset_index(name='total')

fig = px.bar(df_day,
             x="total",
             y="day",
             color="state",
             orientation='h', # Horizontal looks better for long State names
             text_auto='.2s', # Adds shortened numbers (like 12k) on the bars
             title="Total Accidents Day and State")

fig.update_layout(
    yaxis={'categoryorder':'total ascending'},
    hovermode="y unified" # Shows all category data when you hover over a state
)
fig.show()

In [8]:
# which months are most dangerous?
# Welche Monate sind am gefährlichsten?

# 1. Group the data
df_month = df.groupby(['month', 'state', 'category']).size().reset_index(name='total_month')

# 2. Use 'path' for the levels and 'values' for the counts
fig = px.sunburst(
    df_month,
    path=['month', 'state', 'category'],
    values='total_month',
    color='total_month',
    color_continuous_scale=['#f1c21b', '#ff832b', '#da1e28'],
    title="Monthly Accident Distribution"
)

fig.update_layout(margin=dict(t=50, l=0, r=0, b=0))
fig.show()

In [ ]:
#Section 3 —  Weather analysis
#Schritt 3 — Analyse der Wetterbedingungen

# Step 1 — Load data - 2021
# Schritt 1 — Daten laden-2021

weather_2021 = pd.read_csv(os.path.join(WEATHER_DIR, 'combined_2021_hourly_weather_data.csv'), low_memory=False)


In [10]:
weather_2022 = pd.read_csv(os.path.join(WEATHER_DIR, 'combined_2022_hourly_weather_data.csv'), low_memory=False)

In [11]:
weather_2023 = pd.read_csv(os.path.join(WEATHER_DIR, 'combined_2023_hourly_weather_data.csv'), low_memory=False)

In [12]:
# Combine all weather years
# Zusammenführung aller Wetterjahre

weather = pd.concat([weather_2021, weather_2022, weather_2023], ignore_index=True)
print(f"✅ Total weather rows: {len(weather)}")

✅ Total weather rows: 18332424


In [13]:
# Step 2 — Fix timezone
# Schritt 2 — Zeitzone anpassen
# Convert timezone UTC → Berlin
# Zeitzone von UTC in Berliner Zeit umwandeln

weather['date'] = pd.to_datetime(weather['date'], utc=True)
weather['date'] = weather['date'].dt.tz_convert('Europe/Berlin')

In [14]:
# Extract date and hour for joining
weather['join_date'] = weather['date'].dt.strftime('%Y-%m-%d')
weather['join_hour'] = weather['date'].dt.hour

# Round coordinates
weather['lat_round'] = weather['latitude'].round(2)
weather['lon_round'] = weather['longitude'].round(2)

print(f"✅ Timezone converted!")
print(weather[['date', 'join_date', 'join_hour', 'lat_round', 'lon_round']].head())

✅ Timezone converted!
                       date   join_date  join_hour  lat_round  lon_round
0 2021-07-02 00:00:00+02:00  2021-07-02          0      52.43      13.34
1 2021-07-02 01:00:00+02:00  2021-07-02          1      52.43      13.34
2 2021-07-02 02:00:00+02:00  2021-07-02          2      52.43      13.34
3 2021-07-02 03:00:00+02:00  2021-07-02          3      52.43      13.34
4 2021-07-02 04:00:00+02:00  2021-07-02          4      52.43      13.34


In [15]:
# Step 3 - Prepare accidents for joining
# Schritt 3 - Unfalldaten für den Join vorbereiten

df['join_date'] = pd.to_datetime(df['estimated_date']).dt.strftime('%Y-%m-%d')
df['join_hour'] = df['hour'].astype(int)
df['lat_round'] = df['latitude'].round(2)
df['lon_round'] = df['longitude'].round(2)

print(df[['estimated_date', 'join_date',
                  'join_hour', 'lat_round',
                  'lon_round']].head())

  estimated_date   join_date  join_hour  lat_round  lon_round
0     2021-03-01  2021-03-01          7      54.77       8.75
1     2021-06-01  2021-06-01         15      54.32      10.12
2     2021-06-03  2021-06-03         13      53.90       9.41
3     2021-05-03  2021-05-03         11      53.81      10.75
4     2021-05-01  2021-05-01         15      54.10      10.79


In [17]:
# Step 4 — Merge using DuckDB SQL
# Schritt 4 — Daten mit DuckDB SQL zusammenführen

merged = duckdb.query("""
    SELECT
        d.*,
        w.temperature_2m,
        w.rain,
        w.snowfall,
        w.wind_speed_10m,
        w.wind_direction_10m,
        w.weather_code AS weather_code_hourly,
        w.cloud_cover,
        w.shortwave_radiation
    FROM df d
    LEFT JOIN weather w
        ON d.join_date = w.join_date
        AND d.join_hour = w.join_hour
        AND d.lat_round = w.lat_round
        AND d.lon_round = w.lon_round
""").df()

print(f"\n✅ Merged rows: {len(merged)}")
print(f"✅ Columns: {merged.columns.tolist()}")

# Check how many accidents got weather data
matched = merged['rain'].notna().sum()
print(f"✅ Accidents with weather data: {matched} ({matched/len(merged)*100:.1f}%)")

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))


✅ Merged rows: 891257
✅ Columns: ['state', 'government_district', 'district', 'municipality', 'official_admin_code', 'accident_uid', 'year', 'month', 'hour', 'day', 'category', 'kind', 'type', 'light', 'road_surface', 'goods_vehicle_involved', 'passenger_car_involved', 'motorcycle_involved', 'bike_involved', 'passenger_involved', 'other_involved', 'utm_easting', 'utm_northing', 'longitude', 'latitude', 'month_num', 'estimated_date', 'datetime', 'join_date', 'join_hour', 'lat_round', 'lon_round', 'temperature_2m', 'rain', 'snowfall', 'wind_speed_10m', 'wind_direction_10m', 'weather_code_hourly', 'cloud_cover', 'shortwave_radiation']
✅ Accidents with weather data: 888414 (99.7%)


In [40]:
# Save merged dataset
merged.to_csv('merged_accidents_weather.csv', index=False)

In [18]:
# WMO Weather Code descriptions
# WMO Wettercodes Beschreibungen
weather_code_map = {
    0: 'Clear sky',
    1: 'Mainly clear',
    2: 'Partly cloudy',
    3: 'Overcast',
    45: 'Fog',
    48: 'Icy fog',
    51: 'Light drizzle',
    53: 'Moderate drizzle',
    55: 'Heavy drizzle',
    61: 'Slight rain',
    63: 'Moderate rain',
    65: 'Heavy rain',
    71: 'Slight snow',
    73: 'Moderate snow',
    75: 'Heavy snow',
    77: 'Snow grains',
    80: 'Slight showers',
    81: 'Moderate showers',
    82: 'Heavy showers',
    85: 'Slight snow showers',
    86: 'Heavy snow showers',
    95: 'Thunderstorm',
    96: 'Thunderstorm with hail',
    99: 'Thunderstorm heavy hail'
}

# Add to merged dataframe
merged['weather_description'] = merged['weather_code_hourly'].map(weather_code_map).fillna('Unknown')

print(merged[['weather_code_hourly', 'weather_description']].head(10))

   weather_code_hourly weather_description
0                  2.0       Partly cloudy
1                  2.0       Partly cloudy
2                 53.0    Moderate drizzle
3                  1.0        Mainly clear
4                  0.0           Clear sky
5                  0.0           Clear sky
6                 51.0       Light drizzle
7                  3.0            Overcast
8                  0.0           Clear sky
9                  3.0            Overcast


In [ ]:
# Section 5 — Location map

# Map uses 50,000 sampled points for performance
# Full dataset (891,257 rows) used for statistical analysis
# Sample is statistically representative of full data

# Karte verwendet 50.000 Stichprobenpunkte für Performance
# Vollständiger Datensatz für statistische Analyse verwendet

df_map = merged.sample(n=50000, random_state=42)
df_map['temperature_2m'] = df_map['temperature_2m'].round(1)
df_map['wind_speed_10m'] = df_map['wind_speed_10m'].round(1)

fig = px.scatter_map(
    df_map,
    lat='lat_round',
    lon='lon_round',
    color='category',
    color_discrete_map={
        'killed': '#da1e28',
        'seriously injury': '#ff832b',
        'slightly injury': '#f1c21b'
    },
    hover_data={
        # Location
        'state': True,
        'district': True,
        # Time
        'join_date': True,
        'join_hour': True,
        # Weather — readable
        'weather_description': True,
        'rain': True,
        'temperature_2m': True,
        'wind_speed_10m': True,
        'cloud_cover': True,
        # Hide numbers
        'lat_round': False,
        'lon_round': False,
        'weather_code_hourly': False,
        'latitude': False,
        'longitude': False,
    },
    labels={
        'weather_description': '🌤️ Weather',
        'rain': '🌧️ Rain (mm)',
        'temperature_2m': '🌡️ Temp (°C)',
        'wind_speed_10m': '💨 Wind (km/h)',
        'cloud_cover': '☁️ Cloud Cover (%)',
        'join_date': '📅 Date',
        'join_hour': '⏰ Hour',
        'state': '📍 State',
        'district': '🏘️ District',
        'category': '⚠️ Severity'
    },
    zoom=5,
    map_style='open-street-map',
    title='🇩🇪 Accident Locations with Weather Conditions 2021-2023<br><sub>Hover over any point to see details</sub>'
)

fig.update_layout(
    height=700,
    margin=dict(t=80, l=0, r=0, b=0)
)

fig.write_html('accidents_map.html')

import webbrowser, os
webbrowser.open(f'file://{os.path.abspath("accidents_map.html")}')

True

In [21]:
# Section 6 — Gemma 4 AI Analysis
# Abschnitt 6 — Gemma 4 KI-Analyse
# ============================================

# 8.1 — Test connection
# 8.2 — Helper Functions
# 8.3 — Ask about Weather Impact Analysis
# 8.4 — Ask about Location Analysis
# 8.5 — Ask about Time Pattern Analysis
# 8.6 — Final Recommendations

In [22]:
# 8.1 — Test Gemma 4 connection
# Gemma 4 Verbindung testen
response = requests.post(
    "http://localhost:11434/api/generate",
    json={
        "model": "gemma4:e4b",
        "prompt": "Say hello in English and German in one sentence.",
        "stream": False
    }
)

if response.status_code == 200:
    print("✅ Gemma 4 connected successfully!")
    print(response.json()["response"])
else:
    print(f"❌ Error: {response.status_code}")

✅ Gemma 4 connected successfully!
Hello in English and German is, "Hello, ich sage hallo."


In [23]:
# 8.2 — Helper Functions
# Hilfsfunktionen


def ask_gemma(prompt):
    """Send question to Gemma 4 via Ollama"""
    response = requests.post(
        "http://localhost:11434/api/generate",
        json={
            "model": "gemma4:e4b",
            "prompt": prompt,
            "stream": False
        }
    )
    return response.json()["response"]


def ask_and_save(prompt, filename):
    """Ask Gemma and save response — loads from file if already asked"""
    filepath = os.path.join(BASE_DIR, 'gemma_responses', filename)

    # Create folder if not exists
    os.makedirs(os.path.join(BASE_DIR, 'gemma_responses'), exist_ok=True)

    # Load if already saved
    if os.path.exists(filepath):
        print(f"✅ Loading saved: {filename}")
        with open(filepath, 'r') as f:
            return f.read()

    # Ask Gemma
    print(f"🤖 Asking Gemma 4... (may take 5 minutes)")
    print(f"💾 Will save to: {filename}")
    response = ask_gemma(prompt)

    # Save
    with open(filepath, 'w') as f:
        f.write(response)
    print(f"✅ Saved!")

    return response

In [24]:
# ---- 8.3 Weather Impact Analysis ----
# Wettereinfluss Analyse

weather_stats = merged.groupby('category').agg(
    total=('category', 'count'),
    avg_rain=('rain', 'mean'),
    avg_temp=('temperature_2m', 'mean'),
    avg_wind=('wind_speed_10m', 'mean'),
    avg_cloud=('cloud_cover', 'mean'),
    rainy_pct=('rain', lambda x: (x > 0).mean() * 100)
).round(2)

response_weather = ask_and_save(f"""
You are a road safety expert analyzing German accident data 2021-2023.

Weather conditions per accident severity:
{weather_stats.to_string()}

Total accidents: {len(merged):,}
Fatal accidents: {len(merged[merged['category']=='killed']):,}

Analyze:
1. What weather conditions are most associated with fatal accidents?
2. Why do most accidents happen in mild/dry conditions?
3. Role of wind speed in accident severity?
4. Recommendations to reduce fatal accidents?

Be concise. Answer in English then German.
""", "gemma_weather.txt")

print("🌤️ Weather Analysis / Wetteranalyse")
print("=" * 50)
print(response_weather)

🤖 Asking Gemma 4... (may take 5 minutes)
💾 Will save to: gemma_weather.txt
✅ Saved!
🌤️ Weather Analysis / Wetteranalyse
**English Analysis (Road Safety Expert)**

Based on the analysis of German accident data (2021–2023), the following insights regarding weather and severity are presented:

**1. Weather Conditions Associated with Fatal Accidents**
The data shows that fatalities ($\text{Killed}$) are most frequently associated with **moderate and non-extreme** weather conditions. The average rainfall (0.11) and temperature (12.41°C) suggest that weather alone is not the primary cause. The specific average wind speed (12.44 km/h) and cloud cover (67.30%) indicate that predictable, yet suboptimal, conditions contribute significantly to high-severity incidents.

**2. Why Most Accidents Occur in Mild/Dry Conditions**
The overwhelming volume of minor injuries ($\text{Slightly Injury}$) suggests that the primary cause of the vast majority of accidents is **human error** (e.g., distraction, fa

In [25]:
# ---- 8.4 Location Analysis ----
location_stats = merged.groupby('state').agg(
    total=('category', 'count'),
    fatal=('category', lambda x: (x == 'killed').sum()),
    fatal_pct=('category', lambda x: (x == 'killed').mean() * 100)
).round(2).sort_values('total', ascending=False)

response_location = ask_and_save(f"""
You are a road safety expert analyzing German accident data 2021-2023.

Accidents by state (sorted by total):
{location_stats.to_string()}

Analyze:
1. Why does Nordrhein-Westfalen have the most accidents?
2. Which states have highest fatal accident percentage?
3. What regional factors explain these differences?

Be concise. Answer in English then German.
""", "gemma_location.txt")

print("📍 Location Analysis / Standortanalyse")
print("=" * 50)
print(response_location)

🤖 Asking Gemma 4... (may take 5 minutes)
💾 Will save to: gemma_location.txt
✅ Saved!
📍 Location Analysis / Standortanalyse
***English Analysis***

**1. Why does Nordrhein-Westfalen have the most accidents?**
NRW's high accident volume is primarily due to a combination of massive population density, its function as Germany's largest economic and demographic engine (Ruhr area, etc.), and a complex, highly trafficked road infrastructure. The absolute number of accidents correlates strongly with sheer traffic volume ($\text{Unfallaufkommen}$), making it a statistical inevitability regardless of local risk factors.

**2. Which states have the highest fatal accident percentage?**
The states with the highest fatal accident percentages are:
1. Sachsen-Anhalt (1.61%)
2. Brandenburg (1.24%)
3. Baden-Württemberg (1.09%)
4. Niedersachsen (1.18%)

These high percentages suggest that, relative to their total traffic volume, the accident risk severity or specific high-risk driving patterns are more p

In [26]:
# ---- 8.5 Time Pattern Analysis ----
time_stats = merged.groupby(['day', 'hour'])['category'].apply(
    lambda x: (x == 'killed').sum()
).reset_index(name='fatal_count')

most_dangerous = time_stats.nlargest(5, 'fatal_count')

# All accidents by day
all_by_day = merged.groupby('day').size().reset_index(name='total')

# Fatal by day
fatal_by_day = merged[merged['category']=='killed'].groupby('day').size().reset_index(name='fatal')

# Combine
day_analysis = all_by_day.merge(fatal_by_day, on='day')
day_analysis['fatal_pct'] = (day_analysis['fatal'] / day_analysis['total'] * 100).round(2)

response_time = ask_and_save(f"""
You are a road safety expert analyzing German accident data 2021-2023.

Accidents by day of week:
{day_analysis.sort_values('total', ascending=False).to_string()}

Top 5 most deadly day/hour combinations:
{most_dangerous.to_string()}

Key observations:
- Friday has MOST accidents (146,410) but lower fatal rate (0.82%)
- Sunday has FEWEST accidents (81,594) but HIGHEST fatal rate (1.27%)
- Monday 15:00 is single most deadly hour (112 deaths)

Analyze:
1. Why does Sunday have highest fatal rate despite fewer accidents?
2. Role of drunk/impaired driving on weekends?
3. Why is Monday 15:00 most deadly single hour?
4. How can authorities target Sunday and Monday dangers differently?

Be concise. Answer in English then German.
""", "gemma_time.txt")

print("⏰ Time Analysis / Zeitanalyse")
print("=" * 50)
print(response_time)

🤖 Asking Gemma 4... (may take 5 minutes)
💾 Will save to: gemma_time.txt
✅ Saved!
⏰ Time Analysis / Zeitanalyse
**English Analysis**

**1. Why does Sunday have the highest fatal rate despite fewer accidents?**
The high fatal rate suggests a *severity bias*. The accidents that do occur on Sundays are disproportionately critical. Possible causes include:
*   **Fatigue and Planning:** Drivers may be traveling long distances, returning from multi-day activities, or operating under increased fatigue, leading to major, single-vehicle critical crashes.
*   **Behavioral Change:** The demographic profile driving on Sundays might include more risk-taking travelers or those who are operating vehicles outside their normal routine, increasing the lethality of any incident.

**2. Role of drunk/impaired driving on weekends?**
Weekends (especially Saturday) are historically linked to elevated levels of impairment.
*   **Social Factor:** Increased social gathering and consumption (alcohol, substances) e

In [27]:
# ---- 8.6 Final Recommendations ----
response_recommendations = ask_and_save(f"""
You are a road safety expert for Germany's Vision Zero initiative.

Summary of our analysis of {len(merged):,} accidents 2021-2023:

LOCATION:
- Most accidents: Nordrhein-Westfalen (density/urban problem)
- Highest fatal rate: Sachsen-Anhalt 1.61%, Brandenburg 1.24% (rural problem)

TIME PATTERNS:
- Most accidents: Friday (146,410 total)
- Most DEADLY day: Sunday (1.27% fatal rate) — suspected drunk driving
- Most deadly single hour: Monday 15:00 (112 deaths)
- Weekend fatal rate much higher than weekdays

WEATHER:
- Cold temperatures increase fatal risk
- Wind speed not a significant factor
- Most accidents happen in mild/dry conditions (high traffic volume)

FATAL ACCIDENTS:
- Total: {len(merged[merged['category']=='killed']):,}
- Rate: {len(merged[merged['category']=='killed'])/len(merged)*100:.1f}%

Based on this complete analysis, provide:
1. Top 5 actionable recommendations for German road safety
2. Different strategies for West vs East Germany
3. How to target Sunday drunk driving specifically
4. How to reduce Monday afternoon accidents
5. How does this align with Vision Zero goals?

Be specific and data-driven. Answer in English then German.
""", "gemma_recommendations.txt")

print("💡 Recommendations / Empfehlungen")
print("=" * 50)
print(response_recommendations)

🤖 Asking Gemma 4... (may take 5 minutes)
💾 Will save to: gemma_recommendations.txt
✅ Saved!
💡 Recommendations / Empfehlungen
## 🇩🇪 Road Safety Expert Report: Data-Driven Interventions for Vision Zero

Based on the analysis of 891,257 accidents between 2021 and 2023, the pattern is clear: risk is not static. It is concentrated in specific locations, time windows, and behavioral states. Our interventions must therefore be highly targeted.

---

### English Response

#### 1. Top 5 Actionable Recommendations for German Road Safety

These recommendations prioritize systemic change over merely punitive measures.

**R1. Mandatory "High-Risk Behavior" Intervention Zones (HRIZ):**
*   **Action:** Implement advanced speed monitoring and driver behavior intervention (e.g., automated warnings, temporary point deductions) in zones identified by historical data as high-risk (e.g., specific intersections, stretch of roads in Nordrhein-Westfalen).
*   **Data Link:** Addresses the high accident density

# 🔍 Key Insights / Wichtigste Erkenntnisse

## 🌤️ Weather Impact
- Most accidents happen in **mild/dry conditions** — not bad weather
- **Cold temperature** (avg 12.4°C) correlates with fatal accidents
- **Wind speed** has minimal effect on accident severity
- Rain percentage only marginally higher for fatal accidents

## 📍 Location — Two Different Problems
- **West Germany** (NRW, Bayern): Highest accident COUNT
  → Population density + urban complexity
- **East Germany** (Sachsen-Anhalt 1.61%, Brandenburg 1.24%): Highest FATAL rate
  → Rural infrastructure + higher speed roads

## ⏰ Time Patterns — The Surprising Truth
- ❌ Everyone thinks Friday is most dangerous
- ✅ **Sunday is actually most deadly** (1.27% fatal rate)
  → Suspected drunk/impaired driving
  → Leisure drivers less focused than commuters
- **Monday 15:00** is single most deadly hour (112 deaths)
  → "Sunday Scaries" stress + fatigue effect

## 🚗 Scale of Analysis
- **782,642** accidents analyzed
- **3 years** of data (2021-2023)
- **99.7%** weather data matched
- **7,952** fatal accidents

## 💡 Recommendations (Vision Zero)
- Different strategies needed for West vs East Germany
- Sunday: Alcohol intervention + saturation policing
- Monday: Fatigue awareness campaigns 14:00-17:00
- NRW: Engineering solutions (infrastructure)
- East: Behavioral + enforcement solutions